### PR TITLE
[BEAM-8811] Upgrade Beam pipeline diagrams in docs

### PR DESCRIPTION
**Pull request by soyrice**
R: @rosetn, @aaltay
CC: @davidyan74, @kennknowles, @pabloem 

New wordcount pipeline diagram and new instructional diagrams in programming guide/"Design your pipeline" page.

The new diagrams represent:
- PTransforms and PCollections are both as nodes in directed acyclic graphs.
- PCollections functionally, as inputs and outputs for PTransforms.
- The control flow with edges, rather than objects in the programming model.

For source control, the diagrams are SVGs, so you can download the images directly and modify them in graphics editor.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by soyrice** [comment]
Staging links:
http://apache-beam-website-pull-requests.storage.googleapis.com/10200/get-started/wordcount-example/index.html
http://apache-beam-website-pull-requests.storage.googleapis.com/10200/documentation/programming-guide/index.html
http://apache-beam-website-pull-requests.storage.googleapis.com/10200/documentation/pipelines/design-your-pipeline/index.html

**Comment by pabloem** [comment]
My bad if we cleared this during review, but an Aggregation is also a PTransform. The diagrams seem to imply that they're different. Maybe an 'Aggregating PTransform', or a 'Grouping PTransform', vs a 'per-element PTransform'.

**Comment by soyrice** [comment]
> My bad if we cleared this during review, but an Aggregation is also a PTransform. The diagrams seem to imply that they're different. Maybe an 'Aggregating PTransform', or a 'Grouping PTransform', vs a 'per-element PTransform'.

Good catch - thanks. I like "Aggregating PTransform" since it's also consistent with the language in the transform catalog. Will update the legends accordingly.

**Comment by soyrice** [comment]
> > My bad if we cleared this during review, but an Aggregation is also a PTransform. The diagrams seem to imply that they're different. Maybe an 'Aggregating PTransform', or a 'Grouping PTransform', vs a 'per-element PTransform'.
> 
> Good catch - thanks. I like "Aggregating PTransform" since it's also consistent with the language in the transform catalog. Will update the legends accordingly.

Done.

**Comment by pabloem** [comment]
Thanks Cyrus. It looks like http://apache-beam-website-pull-requests.storage.googleapis.com/10200/get-started/wordcount-example/index.html has a figure that needs updating.

**Comment by soyrice** [comment]
> Thanks Cyrus. It looks like http://apache-beam-website-pull-requests.storage.googleapis.com/10200/get-started/wordcount-example/index.html has a figure that needs updating.

Done.

**Comment by pabloem** [comment]
Very cool having the SVG. RAT must be failing because the SVG files don't have the Apache license header. Other than that, LGTM.

**Comment by soyrice** [comment]
> Very cool having the SVG. RAT must be failing because the SVG files don't have the Apache license header. Other than that, LGTM.


> Very cool having the SVG. RAT must be failing because the SVG files don't have the Apache license header. Other than that, LGTM.

Looks like it. Should be fixed now but I'll keep tabs on it in case this next build fails.

**Comment by aaltay** [comment]
LGTM. 

@rosetn - Could you review this please?

File: website/src/documentation/programming-guide.md [hunk 1]
```diff
@@ -512,11 +510,11 @@ you can chain transforms to create a sequential pipeline, like this one:
...
-The resulting workflow graph of the above pipeline looks like this.
+The graph of this pipeline looks like the following:
 
-![This linear pipeline starts with one input collection, sequentially applies
+<br>![This linear pipeline starts with one input collection, sequentially applies
   three transforms, and ends with one output collection.](
-  {{ "/images/design-your-pipeline-linear.png" | prepend: site.baseurl }})
+  {{ "/images/design-your-pipeline-linear.svg" | prepend: site.baseurl }})
 
 *Figure: A linear pipeline with three sequential transforms.*
```
Comment (rosetn) [inline website/src/documentation/programming-guide.md]: Add figure number

File: website/src/documentation/programming-guide.md [hunk 2]
```diff
@@ -536,11 +534,11 @@ a branching pipeline, like so:
...
-The resulting workflow graph from the branching pipeline above looks like this.
+The graph of this branching pipeline looks like the following:
 
-![This pipeline applies two transforms to a single input collection. Each
+<br>![This pipeline applies two transforms to a single input collection. Each
   transform produces an output collection.](
-  {{ "/images/design-your-pipeline-multiple-pcollections.png" | prepend: site.baseurl }})
+  {{ "/images/design-your-pipeline-multiple-pcollections.svg" | prepend: site.baseurl }})
 
 *Figure: A branching pipeline. Two transforms are applied to a single
```
Comment (rosetn) [inline website/src/documentation/programming-guide.md]: Add figure number

File: website/src/images/design-your-pipeline-flatten.svg [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+-->
+<svg width="1094px" height="266px" viewBox="0 0 1094 266" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <title>A pipeline that merges two collections into one collection with the Flatten transform.</title>
```
[the remark this answers is not in the document]
Comment (soyrice) [reply website/src/images/design-your-pipeline-flatten.svg]: The alt attributes are specified in the markdown syntax for the  tag, when the diagrams are inserted into the programming-guide.md and design-your-pipeline.md docs. I kept the same alt text as the original diagrams. Do you think they could use improvements/we should improve them in this PR?

**Comment by soyrice** [comment]
> These diagrams are great! They'll definitely help users grasp concepts more easily. I have a few overall recommendations to improve these figures:
> 
> 1. Add an alt attribute to each of the images for accessibility purposes
> 2. Resize the figures so that on a given page, the font on the figures are the same size. e.g. the sizing on "Figure: A linear pipeline with three sequential transforms." looks appropriate, but you'll need to resize the figure below it

I resized all of the figures that were bigger than the "linear pipeline with three sequential transforms" one.

What do you think we should do about figures that are smaller, like the wordcount example diagram? If I scale these diagrams up, then they're too long to fit on the page. I was thinking of leaving them in the smaller size, but I could also re-draw them vertically so that we could increase the size of the font/diagram components without overflowing. 

It wouldn't take too long to re-draw the diagrams vertically, but one drawback of long, vertical diagrams is they'd take up a lot of space. Although maybe there's a way to wrap the text around the diagram so that it fits neatly in the page...

File: website/src/documentation/programming-guide.md (hunk 1 above)
Comment (soyrice) [reply website/src/documentation/programming-guide.md]: Done.

File: website/src/documentation/programming-guide.md (hunk 2 above)
Comment (soyrice) [reply website/src/documentation/programming-guide.md]: Done.

**Comment by rosetn** [comment]
Your changes look good!

A wrapped flowchart might be a good compromise on web page space and text size readability here. 

The longer figures can be split up, maintaining the left to right reading pattern e.g. https://wcs.smartdraw.com/flowchart/img/wrap-flowchart.jpg. Or you can do the same thing vertically e.g. https://www.researchgate.net/figure/s-a-flowchart-for-wrapping-process-that-has-been-clearly-done-after-the-host-and-the_fig5_320295347

**Comment by soyrice** [comment]
> Your changes look good!
> 
> A wrapped flowchart might be a good compromise on web page space and text size readability here.
> 
> The longer figures can be split up, maintaining the left to right reading pattern e.g. https://wcs.smartdraw.com/flowchart/img/wrap-flowchart.jpg. Or you can do the same thing vertically e.g. https://www.researchgate.net/figure/s-a-flowchart-for-wrapping-process-that-has-been-clearly-done-after-the-host-and-the_fig5_320295347

Done.

**Pull request merged**